ext/LegendGeSimExt.jl

```diff
@@ -0,0 +1,567 @@
+and create a .dat file (unformatted stream of Float32) to be later used when calling fieldgen().
+"""
+function LegendGeSim.impurity_density_model(meta::PropDict, crystal_metadata::PropDict, fit_param::Vector{Float64}, ::SiggenSimulator)   
+
+    a,b,c,tau = fit_param 
+
+    T = Float32
+    dist = T.(crystal_metadata.impurity_measurements.distance_from_seed_end_mm)
+    # sometimes might not have measurements at seed end? but last measurement can be taken as crystal length
+    L = dist[end] #mm
+
+    # points in crystal axis from 0 to L with step of 1 mm
+    # always put 200 points to make all .dat files same length
+    imp = zeros(Float32, 200) # impurities in 10^10 cm^-3
+
+    # length is L+1 including both ends
+    Npoints = Int(ceil(L))+1 # step of 1 mm
+    zpoints = LinRange(0, L, Npoints) 
+    for i in 1:Npoints
+        # divide by 10 to convert from 1e9 e/cm^3 to 1e10 e/cm^3
+        # invert the order of the values because with siggen z=0 is tail (dirtiest part)
+        # i.e. z = L in our system (Mirion measurements)
+        imp[Npoints-i+1] = -(a + b * zpoints[i] + c * exp((zpoints[i] - L)/tau)) / 10.
+    end
+
+    crystal_name = first(meta.name, length(meta.name)-1)
+    imp_filename = joinpath("cache", crystal_name*".dat")
+	open(imp_filename, "w") do io
+		write(io, imp)
+	end    
+    @info "Impurity profile for siggen saved in $(imp_filename)"
+
+    # return name of impurity profile file and offset
+    # offset from seed end
+    det_z0 = crystal_metadata.slices[Symbol(meta.production.slice)].detector_offset_in_mm    
+
+    # siggen offset is from tail i.e. at L offset = 0
+    imp_filename, L - det_z0
+end
+
+"""
+Simulation method: siggen
+"""
+
+
+
+function LegendGeSim.simulate_waveforms(stp_events::Table, detector::SigGenSetup, ::SiggenSimulator)
+    T = Float32 # This should be somehow defined and be passed properly
+    @info("~.~.~ Siggen")
+
+    nevents = length(stp_events)
+    wf_array = Array{RDWaveform}(undef, nevents)
+    for i in 1:nevents
+        # println("$i/$nevents")
+        if(i % 100  == 0) println("$i/$nevents") end
+        signal::Vector{Float32} = simulate_signal(stp_events[i].pos, stp_events[i].edep, detector)
+        time = range(T(0)u"ns", step = detector.step_time_out*u"ns", length = length(signal))
+
+        # push!(waveforms, signal)
+        wf_array[i] = RDWaveform(time, signal)
+    end
+
+    ## Oliver said this should be the more efficient way of doing it
+    ## as opposed to an array of separate RDWaveform instances
+    # Δt = setup.step_time_out*u"ns"
+    # t_end = length(waveforms[1]) * Δt
+    # times = fill(0u"ns" : Δt : t_end, length(waveforms))
+    # ArrayOfRDWaveforms((times, VectorOfSimilarVectors(waveforms)))
+
+    waveforms = ArrayOfRDWaveforms(wf_array)
+    # why am I doing this?
+    waveforms = ArrayOfRDWaveforms((waveforms.time, VectorOfSimilarVectors(waveforms.signal)))    
```

Comment:
@oschulz ?

Reply:
I am gonna resolve this for merging purposes, but @oschulz should comment on this